File: src/chain-observers/providers/reblog-provider.ts
import { TAccountName } from "@hiveio/wax";
import { WorkerBeeIterable } from "../../types/iterator";
import { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import { IOperationTransactionPair, OperationClassifier } from "../classifiers/operation-classifier";
import { TProviderEvaluationContext } from "../factories/data-evaluation-context";
import { ProviderBase } from "./provider-base";

export interface IReblogOperation {
  account: TAccountName;
  author: TAccountName;
  permlink: string;
}

export type TReblogProvided<TAccounts extends Array<TAccountName>> = {
  [K in TAccounts[number]]: WorkerBeeIterable<IOperationTransactionPair<IReblogOperation>>;
};

export interface IReblogProviderData<TAccounts extends Array<TAccountName>> {
  reblogs: Partial<TReblogProvided<TAccounts>>;
};

export interface IReblogProviderOptions {
  accounts: TAccountName[];
}

export class ReblogProvider<TAccounts extends Array<TAccountName> = Array<string>> extends ProviderBase<IReblogProviderOptions> {
  public readonly accounts = new Set<string>();

  public pushOptions(options: IReblogProviderOptions): void {
    for(const account of options.accounts)
      this.accounts.add(account);
  }

  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [OperationClassifier]
  }

  public async provide(data: TProviderEvaluationContext): Promise<IReblogProviderData<TAccounts>> {
    const result = {
      reblogs: {}
    } as IReblogProviderData<TAccounts>;

    const accounts = await data.get(OperationClassifier);
    if (accounts.operationsPerType.custom_json_operation)
      for(const operation of accounts.operationsPerType.custom_json_operation) {
        if (operation.operation.id !== "follow")
          continue;
        let json: [string, IReblogOperation];
        try {
          json = JSON.parse(operation.operation.json);
        } catch {
          continue;
        }

        const reblogger = json[1].account;

        if (json[0] !== "reblog" || !reblogger || !this.accounts.has(reblogger))
          continue;

        if (!result.reblogs[reblogger])
          result.reblogs[reblogger] = [];

        result.reblogs[reblogger].push({
          operation: {
            account: reblogger,
            author: json[1].author,
            permlink: json[1].permlink
          },
          transaction: operation.transaction
        });
      }

    for(const account in result.reblogs)
      result.reblogs[account] = new WorkerBeeIterable(result.reblogs[account]);

    return result;
  }
}

File: python/workerbee/chain_observers/providers/reblog_provider.py
"""ReblogProvider — provides reblog operations grouped by account.

Mirrors src/chain-observers/providers/reblog-provider.ts.
"""

from __future__ import annotations

import json
from typing import TYPE_CHECKING

from .._ordered_set import OrderedSet
from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
from ..classifiers.operation_classifier import OperationClassifier
from .provider_base import ProviderBase

if TYPE_CHECKING:
    from ..factories.data_evaluation_context import DataEvaluationContext
    from ..payloads import ReblogOperationPair, ReblogsPayload

ReblogOptions = dict[str, list[str]]


class ReblogProvider(ProviderBase):
    def __init__(self) -> None:
        self.accounts: OrderedSet[str] = OrderedSet()

    def push_options(self, options: ReblogOptions) -> None:
        accounts = options.get("accounts", [])
        if not accounts:
            raise ValueError("accounts must not be empty")
        for account in accounts:
            self.accounts.add(account)

    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        return [OperationClassifier]

    async def provide(self, data: DataEvaluationContext) -> ReblogsPayload:
        result: dict[str, list[ReblogOperationPair]] = {}

        operations = await data.get(OperationClassifier)
        custom_json_ops = operations["operations_per_type"].get("custom_json_operation")
        if custom_json_ops:
            for operation in custom_json_ops:
                if operation["operation"]["id"] != "follow":
                    continue

                json_payload = operation["operation"].get("json")
                if not isinstance(json_payload, str):
                    continue

                try:
                    parsed = json.loads(json_payload)
                except (json.JSONDecodeError, TypeError):
                    continue

                if not isinstance(parsed, list) or len(parsed) < 2 or not isinstance(parsed[1], dict):
                    continue

                payload = parsed[1]
                reblogger = payload.get("account")
                author = payload.get("author")
                permlink = payload.get("permlink")

                if parsed[0] != "reblog" or not isinstance(reblogger, str) or reblogger not in self.accounts:
                    continue

                if not isinstance(author, str) or not isinstance(permlink, str):
                    continue

                if reblogger not in result:
                    result[reblogger] = []

                result[reblogger].append(
                    {
                        "operation": {
                            "account": reblogger,
                            "author": author,
                            "permlink": permlink,
                        },
                        "transaction": operation["transaction"],
                    }
                )

        return {"reblogs": result}
